{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LABS 8 AND 9"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate Fasta Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command line: /data/home/softwares/SPAdes-3.13.1-Linux/bin/spades.py\t-o\t/bigdata/home/5372607/Lab8and9/spades_output\t-1\t/data/Lab8_data/illumina_reads_R1.fastq\t-2\t/data/Lab8_data/illumina_reads_R2.fastq\t--pacbio\t/data/Lab8_data/pacbio_reads.fastq\t-t\t1\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 3.13.1\n",
      "  Python version: 2.7.16\n",
      "  OS: Linux-4.14.146-119.123.amzn2.x86_64-x86_64-with-glibc2.2.5\n",
      "\n",
      "Output dir: /bigdata/home/5372607/Lab8and9/spades_output\n",
      "Mode: read error correction and assembling\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Multi-cell mode (you should set '--sc' flag if input data was obtained with MDA (single-cell) technology or --meta flag if processing metagenomic dataset)\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/data/Lab8_data/illumina_reads_R1.fastq']\n",
      "      right reads: ['/data/Lab8_data/illumina_reads_R2.fastq']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "    Library number: 2, library type: pacbio\n",
      "      left reads: not specified\n",
      "      right reads: not specified\n",
      "      interlaced reads: not specified\n",
      "      single reads: ['/data/Lab8_data/pacbio_reads.fastq']\n",
      "      merged reads: not specified\n",
      "Read error correction parameters:\n",
      "  Iterations: 1\n",
      "  PHRED offset will be auto-detected\n",
      "  Corrected reads will be compressed\n",
      "Assembly parameters:\n",
      "  k: automatic selection based on read length\n",
      "  Repeat resolution is enabled\n",
      "  Mismatch careful mode is turned OFF\n",
      "  MismatchCorrector will be SKIPPED\n",
      "  Coverage cutoff is turned OFF\n",
      "Other parameters:\n",
      "  Dir for temp files: /bigdata/home/5372607/Lab8and9/spades_output/tmp\n",
      "  Threads: 1\n",
      "  Memory limit (in Gb): 124\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /bigdata/home/5372607/Lab8and9/spades_output/spades.log\n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "== Running read error correction tool: /data/home/softwares/SPAdes-3.13.1-Linux/bin/spades-hammer /bigdata/home/5372607/Lab8and9/spades_output/corrected/configs/config.info\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  75)   Starting BayesHammer, built from refs/heads/spades_3.13.1, git revision 9a9d54db2ff9abaac718155bf74c12ec9464e8ca\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  76)   Loading config from /bigdata/home/5372607/Lab8and9/spades_output/corrected/configs/config.info\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (main.cpp                  :  78)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 124 Gb\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (main.cpp                  :  86)   Trying to determine PHRED offset\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (main.cpp                  :  92)   Determined value is 33\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (hammer_tools.cpp          :  36)   Hamming graph threshold tau=1, k=21, subkmer positions = [ 0 10 ]\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (main.cpp                  : 113)   Size of aux. kmer data 24 bytes\n",
      "     === ITERATION 0 begins ===\n",
      "  0:00:00.002     4M / 4M    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 16 files using 1 threads. This might take a while.\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 41.332 Gb\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 4194304\n",
      "  0:00:00.002   580M / 616M  INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /data/Lab8_data/illumina_reads_R1.fastq\n",
      "  0:00:11.987   580M / 616M  INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 108983 reads\n",
      "  0:00:23.645   580M / 616M  INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 217655 reads\n",
      "  0:00:35.516   580M / 616M  INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 326936 reads\n",
      "  0:00:47.712   580M / 616M  INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 437997 reads\n",
      "  0:00:59.894   580M / 616M  INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 548947 reads\n",
      "  0:01:01.277   580M / 616M  INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /data/Lab8_data/illumina_reads_R2.fastq\n",
      "  0:02:02.774   580M / 616M  INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 1116444 reads\n",
      "  0:02:03.672   580M / 616M  INFO   K-mer Splitting          (kmer_data.cpp             : 112)   Total 1125334 reads processed\n",
      "  0:02:03.703     4M / 616M  INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:02:36.234     4M / 616M  INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 372650690 kmers in total.\n",
      "  0:02:36.234     4M / 616M  INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:02:39.116     4M / 616M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:03:55.898   212M / 616M  INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:04:08.857   212M / 616M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 172813200 bytes occupied (3.70992 bits per kmer).\n",
      "  0:04:08.857   212M / 616M  INFO   K-mer Counting           (kmer_data.cpp             : 356)   Arranging kmers in hash map order\n",
      "  0:05:06.897     5G / 5G    INFO    General                 (main.cpp                  : 148)   Clustering Hamming graph.\n",
      "  1:29:58.253     5G / 5G    INFO    General                 (main.cpp                  : 155)   Extracting clusters\n",
      "  1:32:36.405     5G / 18G   INFO    General                 (main.cpp                  : 167)   Clustering done. Total clusters: 257665812\n",
      "  1:32:36.570     2G / 18G   INFO   K-mer Counting           (kmer_data.cpp             : 376)   Collecting K-mer information, this takes a while.\n",
      "  1:32:41.347    11G / 18G   INFO   K-mer Counting           (kmer_data.cpp             : 382)   Processing /data/Lab8_data/illumina_reads_R1.fastq\n",
      "  1:36:05.710    11G / 18G   INFO   K-mer Counting           (kmer_data.cpp             : 382)   Processing /data/Lab8_data/illumina_reads_R2.fastq\n",
      "  1:39:29.673    11G / 18G   INFO   K-mer Counting           (kmer_data.cpp             : 389)   Collection done, postprocessing.\n",
      "  1:39:30.830    11G / 18G   INFO   K-mer Counting           (kmer_data.cpp             : 403)   There are 372650690 kmers in total. Among them 343377334 (92.1446%) are singletons.\n",
      "  1:39:30.830    11G / 18G   INFO    General                 (main.cpp                  : 173)   Subclustering Hamming graph\n",
      "  2:43:26.117    11G / 18G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 649)   Subclustering done. Total 27212 non-read kmers were generated.\n",
      "  2:43:26.117    11G / 18G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 650)   Subclustering statistics:\n",
      "  2:43:26.117    11G / 18G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 651)     Total singleton hamming clusters: 235632830. Among them 1862576 (0.790457%) are good\n",
      "  2:43:26.117    11G / 18G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 652)     Total singleton subclusters: 17648. Among them 17274 (97.8808%) are good\n",
      "  2:43:26.117    11G / 18G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 653)     Total non-singleton subcluster centers: 22847575. Among them 5230963 (22.895%) are good\n",
      "  2:43:26.117    11G / 18G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 654)     Average size of non-trivial subcluster: 6.0039 kmers\n",
      "  2:43:26.117    11G / 18G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 655)     Average number of sub-clusters per non-singleton cluster: 1.03777\n",
      "  2:43:26.117    11G / 18G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 656)     Total solid k-mers: 7110813\n",
      "  2:43:26.117    11G / 18G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 657)     Substitution probabilities: [4,4]((0.852323,0.0583978,0.035294,0.0539856),(0.0511217,0.871221,0.0306313,0.0470259),(0.0429617,0.0278175,0.871869,0.057352),(0.0520827,0.0339559,0.0577655,0.856196))\n",
      "  2:43:26.424    11G / 18G   INFO    General                 (main.cpp                  : 178)   Finished clustering.\n",
      "  2:43:26.425    11G / 18G   INFO    General                 (main.cpp                  : 197)   Starting solid k-mers expansion in 1 threads.\n",
      "  2:45:47.204    11G / 18G   INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 0 produced 24114 new k-mers.\n",
      "  2:48:06.974    11G / 18G   INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 1 produced 525 new k-mers.\n",
      "  2:50:24.815    11G / 18G   INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 2 produced 0 new k-mers.\n",
      "  2:50:24.815    11G / 18G   INFO    General                 (main.cpp                  : 222)   Solid k-mers finalized\n",
      "  2:50:24.815    11G / 18G   INFO    General                 (hammer_tools.cpp          : 220)   Starting read correction in 1 threads.\n",
      "  2:50:24.815    11G / 18G   INFO    General                 (hammer_tools.cpp          : 233)   Correcting pair of reads: /data/Lab8_data/illumina_reads_R1.fastq and /data/Lab8_data/illumina_reads_R2.fastq\n",
      "  2:50:25.620    11G / 18G   INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 0 of 100000 reads.\n",
      "  3:42:44.563    11G / 18G   INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 0\n",
      "  3:42:44.836    11G / 18G   INFO    General                 (hammer_tools.cpp          : 185)   Written batch 0\n",
      "  3:42:45.586    11G / 18G   INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 1 of 100000 reads.\n",
      "  5:34:04.284    11G / 18G   INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 2\n",
      "  5:34:04.565    11G / 18G   INFO    General                 (hammer_tools.cpp          : 185)   Written batch 2\n",
      "  5:34:05.291    11G / 18G   INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 3 of 100000 reads.\n",
      "  6:31:36.441    11G / 18G   INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 3\n",
      "  6:31:36.723    11G / 18G   INFO    General                 (hammer_tools.cpp          : 185)   Written batch 3\n",
      "  6:31:37.447    11G / 18G   INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 4 of 100000 reads.\n",
      "  7:29:08.019    11G / 18G   INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 4\n",
      "  7:29:08.297    11G / 18G   INFO    General                 (hammer_tools.cpp          : 185)   Written batch 4\n",
      "  7:29:08.749    11G / 18G   INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 5 of 62667 reads.\n",
      "  8:04:22.355    11G / 18G   INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 5\n",
      "  8:04:22.537    11G / 18G   INFO    General                 (hammer_tools.cpp          : 185)   Written batch 5\n",
      "  8:04:22.643    11G / 18G   INFO    General                 (hammer_tools.cpp          : 274)   Correction done. Changed 9637464 bases in 686832 reads.\n",
      "  8:04:22.643    11G / 18G   INFO    General                 (hammer_tools.cpp          : 275)   Failed to correct 447578 bases out of 335616721.\n",
      "  8:04:23.266     8M / 18G   INFO    General                 (main.cpp                  : 255)   Saving corrected dataset description to /bigdata/home/5372607/Lab8and9/spades_output/corrected/corrected.yaml\n",
      "  8:04:23.266     8M / 18G   INFO    General                 (main.cpp                  : 262)   All done. Exiting.\n",
      "\n",
      "== Compressing corrected reads (with gzip)\n",
      "\n",
      "== Dataset description file was created: /bigdata/home/5372607/Lab8and9/spades_output/corrected/corrected.yaml\n",
      "\n",
      "\n",
      "===== Read error correction finished. \n",
      "\n",
      "\n",
      "===== Assembling started.\n",
      "\n",
      "\n",
      "== Running assembler: K21\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /bigdata/home/5372607/Lab8and9/spades_output/K21/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 124 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from refs/heads/spades_3.13.1, git revision 9a9d54db2ff9abaac718155bf74c12ec9464e8ca\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/bigdata/home/5372607/Lab8and9/spades_output/dataset.info) with K=21\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  52)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  59)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  66)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  77)   Converting reads to binary format for library #0 (takes a while)\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  78)   Converting paired reads\n",
      "  0:00:00.498    96M / 132M  INFO    General                 (binary_converter.hpp      :  93)   16384 reads processed\n",
      "  0:00:00.874   120M / 132M  INFO    General                 (binary_converter.hpp      :  93)   32768 reads processed\n",
      "  0:00:01.625   176M / 176M  INFO    General                 (binary_converter.hpp      :  93)   65536 reads processed\n",
      "  0:00:03.121   284M / 284M  INFO    General                 (binary_converter.hpp      :  93)   131072 reads processed\n",
      "  0:00:06.118   504M / 504M  INFO    General                 (binary_converter.hpp      :  93)   262144 reads processed\n",
      "  0:00:12.211   940M / 940M  INFO    General                 (binary_converter.hpp      :  93)   524288 reads processed\n",
      "  0:00:16.165   996M / 996M  INFO    General                 (binary_converter.hpp      : 117)   556092 reads written\n",
      "  0:00:16.536     4M / 996M  INFO    General                 (read_converter.hpp        :  87)   Converting single reads\n",
      "  0:00:16.901   136M / 996M  INFO    General                 (binary_converter.hpp      : 117)   6543 reads written\n",
      "  0:00:16.911     4M / 996M  INFO    General                 (read_converter.hpp        :  95)   Converting merged reads\n",
      "  0:00:17.178   132M / 996M  INFO    General                 (binary_converter.hpp      : 117)   0 reads written\n",
      "  0:00:17.185     4M / 996M  INFO    General                 (construction.cpp          : 111)   Max read length 300\n",
      "  0:00:17.185     4M / 996M  INFO    General                 (construction.cpp          : 117)   Average read length 298.143\n",
      "  0:00:17.185     4M / 996M  INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:17.185     4M / 996M  INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 1 files using 1 threads. This might take a while.\n",
      "  0:00:17.185     4M / 996M  INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:17.185     4M / 996M  INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 41.332 Gb\n",
      "  0:00:17.185     4M / 996M  INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 67108864\n",
      "  0:00:32.394   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 484383 reads\n",
      "  0:00:47.298   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 968718 reads\n",
      "  0:01:02.401   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 1453046 reads\n",
      "  0:01:17.599   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 1937221 reads\n",
      "  0:01:26.932   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 2237454 reads\n",
      "  0:01:26.932   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:01:26.962     4M / 1G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 2237454 reads\n",
      "  0:01:26.962     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:01:36.491     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 142124694 kmers in total.\n",
      "  0:01:36.491     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:01:37.354     4M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:01:37.354     4M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:01:37.354     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 16 files using 1 threads. This might take a while.\n",
      "  0:01:37.354     4M / 1G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:01:37.354     4M / 1G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 41.332 Gb\n",
      "  0:01:37.354     4M / 1G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 4194304\n",
      "  0:01:50.878   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 33510059 kmers\n",
      "  0:02:05.155   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 67047894 kmers\n",
      "  0:02:19.616   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 100577222 kmers\n",
      "  0:02:34.166   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 134106717 kmers\n",
      "  0:02:37.337   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 142124698 kmers\n",
      "  0:02:37.337   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 142124698 kmers.\n",
      "  0:02:37.371     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:02:48.926     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 137374428 kmers in total.\n",
      "  0:02:48.926     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:02:49.744     4M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:03:16.094    64M / 1G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:03:20.211    64M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 63708816 bytes occupied (3.71008 bits per kmer).\n",
      "  0:03:20.323   196M / 1G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:05:02.431   196M / 1G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:05:02.431   196M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
      "  0:05:02.431   196M / 1G    INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
      "  0:05:02.431   196M / 1G    INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n",
      "  0:08:21.696   196M / 1G    INFO   Early tip clipping       (early_simplification.hpp  : 184)   60714991 22-mers were removed by early tip clipper\n",
      "  0:08:21.696   196M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:08:21.696   196M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:10:00.823   680M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 9858295 sequences extracted\n",
      "  0:10:42.907   680M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:11:02.695   680M / 1G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 1 loops collected\n",
      "  0:11:18.207     2G / 3G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:11:18.207     2G / 3G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:11:18.207     2G / 3G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:12:08.355     2G / 3G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 65904104 bytes occupied (3.70965 bits per kmer).\n",
      "  0:12:08.679     3G / 3G    INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:13:48.229     3G / 3G    INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:13:51.611     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 1000000 edges\n",
      "  0:13:54.769     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 2000000 edges\n",
      "  0:13:57.926     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 3000000 edges\n",
      "  0:14:01.141     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 4000000 edges\n",
      "  0:14:04.290     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 5000000 edges\n",
      "  0:14:07.471     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 6000000 edges\n",
      "  0:14:10.602     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 7000000 edges\n",
      "  0:14:13.748     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 8000000 edges\n",
      "  0:14:16.907     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 9000000 edges\n",
      "  0:14:20.034     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 10000000 edges\n",
      "  0:14:23.190     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 11000000 edges\n",
      "  0:14:26.351     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 12000000 edges\n",
      "  0:14:29.520     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 13000000 edges\n",
      "  0:14:32.657     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 14000000 edges\n",
      "  0:14:42.075     3G / 3G    INFO    General                 (construction.cpp          : 464)   Processed 17000000 edges\n",
      "  0:14:51.021     2G / 3G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:14:51.028     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 5\n",
      "  0:14:51.029     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 12\n",
      "  0:14:51.029     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 13. Coverage mad: 5.9304\n",
      "  0:14:51.029     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:14:51.047     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:14:51.094     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:14:51.257     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:14:51.552     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 16\n",
      "  0:14:52.057     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 32\n",
      "  0:14:52.147     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 14.4664. Fitted coverage std. dev: 5.03367\n",
      "  0:14:52.148     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 0.612177\n",
      "  0:14:52.148     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 14\n",
      "  0:14:52.148     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 9\n",
      "  0:14:52.148     2G / 3G    INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 6053702\n",
      "  0:14:52.148     2G / 3G    INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 14.4664\n",
      "  0:14:52.148     2G / 3G    INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 9\n",
      "  0:14:52.148     2G / 3G    INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 0\n",
      "  0:14:52.148     2G / 3G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:14:52.148     2G / 3G    INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:14:52.148     2G / 3G    INFO    General                 (graph_simplification.hpp  : 669)   Flanking coverage based disconnection disabled\n",
      "  0:14:52.148     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:14:54.328     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:14:54.328     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial tip clipper\n",
      "  0:14:59.671     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial tip clipper triggered 15066 times\n",
      "  0:14:59.671     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial ec remover\n",
      "  0:16:23.179     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial ec remover triggered 3816060 times\n",
      "  0:16:23.179     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial isolated edge remover\n",
      "  0:16:24.273     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial isolated edge remover triggered 402 times\n",
      "  0:16:24.771     2G / 3G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:16:24.771     2G / 3G    INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:16:24.771     2G / 3G    INFO    General                 (graph_simplification.hpp  : 641)   Creating parallel br instance\n",
      "  0:16:24.771     2G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:16:24.771     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:16:26.616     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 36989 times\n",
      "  0:16:26.616     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:44:20.414     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 7118 times\n",
      "  0:44:20.414     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:44:23.524     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:44:23.524     2G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:44:23.524     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:44:23.535     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 52 times\n",
      "  0:44:23.535     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:44:23.609     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 4 times\n",
      "  0:44:23.609     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:44:36.860     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 432427 times\n",
      "  0:44:36.860     2G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:44:36.860     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:44:37.198     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 15389 times\n",
      "  0:44:37.198     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:46:29.875     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 226 times\n",
      "  0:46:29.875     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:46:31.608     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 85333 times\n",
      "  0:46:31.608     2G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:46:31.608     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:46:31.667     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 702 times\n",
      "  0:46:31.667     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:46:57.171     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 35 times\n",
      "  0:46:57.171     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:46:59.871     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 127119 times\n",
      "  0:46:59.871     2G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:46:59.871     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:46:59.956     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 744 times\n",
      "  0:46:59.956     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:47:23.585     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 65 times\n",
      "  0:47:23.585     2G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:47:25.314     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 72660 times\n",
      "  0:47:25.314     1G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:47:25.314     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:47:25.356     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 310 times\n",
      "  0:47:25.356     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:47:37.828     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 40 times\n",
      "  0:47:37.828     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:47:38.537     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 38929 times\n",
      "  0:47:38.537     1G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:47:38.537     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:47:38.755     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 158 times\n",
      "  0:47:38.755     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:47:45.526     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 30 times\n",
      "  0:47:45.526     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:47:46.037     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 28848 times\n",
      "  0:47:46.037     1G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:47:46.037     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:47:46.054     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 94 times\n",
      "  0:47:46.054     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:47:50.540     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 28 times\n",
      "  0:47:50.540     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:47:51.089     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 21438 times\n",
      "  0:47:51.089     1G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:47:51.089     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:47:51.102     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 66 times\n",
      "  0:47:51.102     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:47:54.082     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 21 times\n",
      "  0:47:54.082     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:47:54.359     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 15974 times\n",
      "  0:47:54.359     1G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:47:54.359     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:47:54.370     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 43 times\n",
      "  0:47:54.371     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:47:56.503     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 13 times\n",
      "  0:47:56.503     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:47:56.718     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 12509 times\n",
      "  0:47:56.718     1G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:47:56.718     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:47:56.839     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 30 times\n",
      "  0:47:56.839     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:48:07.255     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 21 times\n",
      "  0:48:07.256     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:48:07.437     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:48:07.437     1G / 3G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:48:07.437     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:48:07.438     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:48:07.438     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:48:07.438     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:48:07.438     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:48:07.438     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:48:07.438     1G / 3G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
      "  0:48:07.438     1G / 3G    INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
      "  0:48:07.438     1G / 3G    INFO    General                 (graph_simplification.hpp  : 458)   Disconnection of relatively low covered edges disabled\n",
      "  0:48:07.438     1G / 3G    INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:48:07.438     1G / 3G    INFO    General                 (graph_simplification.hpp  : 641)   Creating parallel br instance\n",
      "  0:48:07.438     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:48:07.554     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:48:07.554     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:48:17.888     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:48:17.888     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:48:18.004     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:48:18.004     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:48:28.590     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:48:28.590     1G / 3G    INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
      "  0:48:28.748     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
      "  0:48:28.843     1G / 3G    INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 12 times\n",
      "  0:48:28.843     1G / 3G    INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
      "  0:48:28.921     1G / 3G    INFO    General                 (simplification.cpp        : 476)   Average coverage = 21.7416\n",
      "  0:48:28.921     1G / 3G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:48:28.921     1G / 3G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /bigdata/home/5372607/Lab8and9/spades_output//K21/simplified_contigs.fasta\n",
      "  0:48:29.562     1G / 3G    INFO    General                 (launch.hpp                : 151)   SPAdes finished\n",
      "  0:48:31.398     4M / 3G    INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 48 minutes 31 seconds\n",
      "Max read length detected as 300\n",
      "Default k-mer sizes were set to [21, 33, 55, 77, 99, 127] because estimated read length (300) is equal to or greater than 250\n",
      "\n",
      "== Running assembler: K33\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /bigdata/home/5372607/Lab8and9/spades_output/K33/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 124 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from refs/heads/spades_3.13.1, git revision 9a9d54db2ff9abaac718155bf74c12ec9464e8ca\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/bigdata/home/5372607/Lab8and9/spades_output/dataset.info) with K=33\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 1\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  52)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  59)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  66)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 300\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 298.143\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 1 files using 1 threads. This might take a while.\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 41.332 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 33554432\n",
      "  0:00:08.562   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 253154 reads\n",
      "  0:00:16.836   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 506298 reads\n",
      "  0:00:25.102   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 759474 reads\n",
      "  0:00:33.367   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 1012551 reads\n",
      "  0:00:41.685   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 1265671 reads\n",
      "  0:00:50.044   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 1518775 reads\n",
      "  0:01:13.719   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 2237454 reads\n",
      "  0:01:13.719   568M / 1G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:01:16.188     4M / 1G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 2237454 reads\n",
      "  0:01:16.188     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:01:32.022     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 166075468 kmers in total.\n",
      "  0:01:32.022     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:01:34.437     4M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:01:34.437     4M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:01:34.437     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 16 files using 1 threads. This might take a while.\n",
      "  0:01:34.437     4M / 1G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:01:34.437     4M / 1G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 41.332 Gb\n",
      "  0:01:34.437     4M / 1G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 2097152\n",
      "  0:01:42.155   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 16757598 kmers\n",
      "  0:01:49.816   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 33517238 kmers\n",
      "  0:01:57.532   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 50279556 kmers\n",
      "  0:02:05.489   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 67038770 kmers\n",
      "  0:02:13.434   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 83803411 kmers\n",
      "  0:02:21.315   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 100562422 kmers\n",
      "  0:02:29.379   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 117320281 kmers\n",
      "  0:02:37.556   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 134079212 kmers\n",
      "  0:02:45.637   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 150837957 kmers\n",
      "  0:02:53.246   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 166075477 kmers\n",
      "  0:02:53.246   580M / 1G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 166075477 kmers.\n",
      "  0:02:53.280     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:03:10.262     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 164797011 kmers in total.\n",
      "  0:03:10.262     4M / 1G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:03:12.183     4M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:03:42.901    76M / 1G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:03:59.030    76M / 1G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 76424568 bytes occupied (3.71 bits per kmer).\n",
      "  0:03:59.162   236M / 1G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:06:03.974   236M / 1G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:06:03.974   236M / 1G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
      "  0:06:03.974   236M / 1G    INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
      "  0:06:03.974   236M / 1G    INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n"
     ]
    }
   ],
   "source": [
    "! spades.py -o spades_output -1 /data/Lab8_data/illumina_reads_R1.fastq -2 /data/Lab8_data/illumina_reads_R2.fastq --pacbio /data/Lab8_data/pacbio_reads.fastq \\-t 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! spades.py -o . -t 1 --restart-from last"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Questions from Lab8 pdf**\n",
    "\n",
    "*Why do we expect short reads to produce a more fragmented assembly than long reads?*\n",
    "\n",
    "Short reads will produce a more fragmented assembly because it increases the likelihood of \"break points\" or regions of non-overlapping reads.\n",
    "\n",
    "*Why does a single-molecule sequencing like PacBio have a higher error rate than Illumina?*\n",
    "\n",
    "PacBio has a higher error rate because there are no other sequences to compare against for validation. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculating Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stats for /data/Lab9_data/contigs.fasta\r\n",
      "sum = 6680581, n = 180, ave = 37114.34, largest = 687571\r\n",
      "N50 = 234690, n = 10\r\n",
      "N60 = 202142, n = 13\r\n",
      "N70 = 154232, n = 17\r\n",
      "N80 = 103366, n = 22\r\n",
      "N90 = 81821, n = 29\r\n",
      "N100 = 128, n = 180\r\n",
      "N_count = 0\r\n",
      "Gaps = 0\r\n"
     ]
    }
   ],
   "source": [
    "! assembly-stats /data/Lab9_data/contigs.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stats for /data/Lab9_data/scaffolds.fasta\r\n",
      "sum = 6680345, n = 176, ave = 37956.51, largest = 687571\r\n",
      "N50 = 234690, n = 10\r\n",
      "N60 = 202142, n = 13\r\n",
      "N70 = 154232, n = 17\r\n",
      "N80 = 103366, n = 22\r\n",
      "N90 = 83515, n = 29\r\n",
      "N100 = 129, n = 176\r\n",
      "N_count = 20\r\n",
      "Gaps = 2\r\n"
     ]
    }
   ],
   "source": [
    "! assembly-stats /data/Lab9_data/scaffolds.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Questions from Lab 9 pdf\n",
    "\n",
    "**Why is N50 is useful statistic to calculate?**\n",
    "\n",
    "According to the wikipedia page referenced in the lab pdf, the N50 statistic allows you to determine the minimum contig sequence length at which 50% or more of the genome is covered. The \"N50 statistic defines assembly quality in terms of contiguity.\"\n",
    "\n",
    "**Why not just list the mean or median contig length?**\n",
    "\n",
    "There might be a skewed distribution lengths, which the mean and median wouldn't capture. \n",
    "\n",
    "**In your assembly, are scaffolds longer than contigs, or are scaffolds approximately equal to contigs?**\n",
    "\n",
    "The scaffolds are approximately equal to the contigs, but I also used the Lab 9 data that we were provided because my Lab 8 results weren't loading properly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Isolate and Analyze 16s Subunit Regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "! rna_hmm3.py -i /data/Lab9_data/contigs.fasta -o output.gff -L /data/Lab9_data/HMM3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "! bedtools getfasta -fi /data/Lab9_data/contigs.fasta -bed output.gff > 16sRNA.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From SeqMatch tool:\n",
    "\n",
    "domain Bacteria \n",
    "\n",
    "phylum \"Proteobacteria\"\n",
    "\n",
    "class Gammaproteobacteria  \n",
    "\n",
    "order Pseudomonadales \n",
    "\n",
    "family Pseudomonadaceae \n",
    "\n",
    "*genus Pseudomonas* \n",
    "\n",
    "I got this result by running the SeqMatch program on the 16s RNA sequences we isolated. Because these genes exist in most or all bacteria, but still have variety, we can use them to identify the homology of the DNA sequence we're analyzing. \n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Species of bacteria within the Pseudomonas genus have become a prevalent pathogen over the last 20 years, causing between 10-20% of the infections incurred in hospitals. Patients with penetrative injuries or breaks in their skin barrier are susceptible to this disease, as well as patients with cystic fibrosis, leukemia, and blood transplants. The bacteria can cause inflammation in various organ systems, leading to severe complications beyond the reason the patient was admitted. There are many antibiotics that have been used to help patients fight off Pseudomonas infections, but the bacteria has also been shown to develop antibiotic resistance. Specifically, Pseudomonas aeruginosa can lead to infections that are hard to get rid of with antibiotics. The primary mechanism of resistance for these bacteria is an immobile biofilm, which imparts resistance to colonies of the bacteria. When comparing Pseudomonas that form biofilms from those that dont, it was found that only 1% of genes have different expression between the two. This shows that there is a defined subset of genes that responsible for Pseudomonas resistance to tobramycin when activated. \n",
    "\n",
    "Sources: \n",
    "Gerald P. Bodey, Ricardo Bolivar, Victor Fainstein, Leena Jadeja, Infections Caused by Pseudomonas aeruginosa, Reviews of Infectious Diseases, Volume 5, Issue 2, March 1983, Pages 279313, https://doi.org/10.1093/clinids/5.2.279\n",
    "\n",
    "Whiteley, M., Bangera, M., Bumgarner, R. et al. Gene expression in Pseudomonas aeruginosa biofilms. Nature 413, 860864 (2001) doi:10.1038/35101627\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
